gaplint.py

```diff
@@ -695,56 +757,270 @@ def _parse_args(kwargs):
+                            [r'(:|>|<)=', r'^\s*=', r'\\=']),
+         WhitespaceOperator('whitespace-mapping', 'W023', r'->'),
+         WhitespaceOperator('whitespace-forward-slash', 'W024', r'\/', 
+                            [r'\\\/']),
+         WhitespaceOperator('whitespace-^', 'W025', r'\^', 
+                            [r'^\s*\^', r'\\\^']),
+         WhitespaceOperator('whitespace-operator-not-equal', 'W026', 
+                            r'<>', [r'^\s*<>']),
+         WhitespaceOperator('whitespace-double-dot', 'W027', r'\.\.', 
+                            [r'\.\.(\.|\))']),
+         UnusedLVarsFunc('unused-local-variables', 'W028')]
+
+_RULE_NAMES_AND_CODES = []
+for rule in RULES:
+    _RULE_NAMES_AND_CODES.append([rule.name, rule.code])
+_RULE_NAMES = [x[0] for x in _RULE_NAMES_AND_CODES] + ['all']
+_RULE_CODES = [x[1] for x in _RULE_NAMES_AND_CODES]
+
+################################################################################
+# suppressions
+################################################################################
+
+def rule_code(rule):
+    '''
+    Takes a rule name or code and returns the rule code.
+    '''
+    global _RULE_NAMES, _RULE_CODES
+    if rule in _RULE_NAMES:
+        return _RULE_CODES[_RULE_NAMES.index(rule)]
+    elif rule in _RULE_CODES:
+        return rule
+    return False
+
+def make_dic(keys, vals): 
+    '''
+    Takes a list of keys and values and returns a dictionary.
+    '''
+    dic = {}
+    n = len(keys)
+    for i in xrange(n):
+        dic[keys[i]] = vals[i]
+    return dic
+
+def make_True_dic(keys):
+    '''
+    Takes a list of keys and returns a dictionary with all keys assigned the
+    value True.
+    '''
+    return make_dic(keys, [True for x in xrange(len(keys))])
+
+def filter_empty_dics(dic): # currently not used - maybe useful later
+    '''
+    Recursive function that takes a dictionary of any dimension and removes 
+    entries whose value is an empty dictionary.
+    '''
+    for key in dic.keys():
+        val = dic[key]
+        if isinstance(val, dict):
+            if val == {}:
+                del dic[key]
+            filter_no_suppression_lines(val)
+
+def is_valid_rule(rule):
+    '''
+    Takes a rule and confirms its validity, returning a bool.
+    '''
+    global _RULE_NAMES, _CODE_NAMES
+    if rule in _RULE_NAMES + _RULE_CODES:
+        return True
+    return False
+
+def get_global_suppressions(lines):
+    '''
+    Takes a list of lines and returns a dictionary with globally suppressed 
+    rules as entries, all assigned the value True.
+    '''
+    global _RULE_CODES
+    pattern1 = re.compile('\s*#')
+    pattern2 = re.compile('\s*$')
+    pattern3 = re.compile('\s*#\s*gaplint:\s*disable\s*=\s*')
+    global_supps = []
+    i = 0
+    n = len(lines)
+    myline = lines[0]
+    while  re.match(pattern1, myline) or re.match(pattern2, myline):
+        match = re.search(pattern3, myline)
```

Comment:
I don't really understand the logic of this, if a line matches any number of whitespace followed by a `#`, then we look for `pattern3`?  Why is this `while` loop required at all?